tedana/metrics/external.py

```diff
@@ -0,0 +1,650 @@
+"""Metrics unrelated to TE-(in)dependence."""
+
+import logging
+import re
+
+import numpy as np
+import pandas as pd
+from scipy import stats
+
+from tedana.stats import fit_model
+
+LGR = logging.getLogger("GENERAL")
+RepLGR = logging.getLogger("REPORT")
+
+
+class RegressError(Exception):
+    """Passes errors that are raised when `validate_extern_regress` fails."""
+
+    pass
+
+
+def load_validate_external_regressors(external_regressors, external_regressor_config, n_time):
+    """
+    Load and validate external regressors and descriptors in dictionary.
+
+    Parameters
+    ----------
+    external_regressors: :obj:`str`
+        Path and name of tsv file that includes external regressor time series
+    external_regressor_config: :obj:`dict`
+        A dictionary with info for fitting external regressors
+        to component time series
+    n_time: :obj:`int`
+        Number of timepoints in the fMRI time series
+
+    Returns
+    -------
+    external_regressors: :obj:`pandas.DataFrame`
+        Each column is a labelled regressor and the number of rows should
+        match the number of timepoints in the fMRI time series
+    external_regressor_config: :obj:`dict`
+        A validated dictionary with info for fitting external regressors
+        to component time series. If regex patterns like '^mot_.*$' are used to define
+        regressor names, this is replaced with a list of the match column names used in
+        external_regressors
+    """
+    try:
+        external_regressors = pd.read_table(external_regressors)
+    except FileNotFoundError:
+        raise ValueError(f"Cannot load tsv file with external regressors: {external_regressors}")
+
+    external_regressor_config = validate_extern_regress(
+        external_regressors, external_regressor_config, n_time
+    )
+
+    return external_regressors, external_regressor_config
+
+
+def validate_extern_regress(external_regressors, external_regressor_config, n_time):
+    """
+    Confirm that provided external regressor dictionary is valid and matches data.
+
+    Checks if expected keys are in external_regressor_config.
+    Checks if any regressor labels in the dictionary are specified in the
+    user-defined external_regressors
+    Checks if the number of time points in the external regressors matches
+    the number of time point in the fMRI data
+
+    Parameters
+    ----------
+    external_regressors : :obj:`pandas.DataFrame`
+        Each column is a labelled regressor and the number of rows should
+        match the number of timepoints in the fMRI time series
+    external_regressor_config : :obj:`dict`
+        Information describing the external regressors and
+        method to use for fitting and statistical tests
+    n_time : :obj:`int`
+        The number of time point in the fMRI time series
```

Comment:
We generally use `n_vols` for this in the rest of the package.

Reply:
I changed `n_time` to `n_vols` everywhere in this file.

Reply:
Thanks!